### PR TITLE
SparseMerkleTree outline

### PR DESCRIPTION
**Pull request by AlCutter**

### CODE REVIEW

File: merkle/map_hasher.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
+package merkle
+
+import (
+	"github.com/google/trillian"
+)
+
+// MapHasher is a specialised TreeHasher which also knows about the set of
+// "null" hashes for the unused sections of a SparseMerkleTree.
+type MapHasher struct {
```
Comment (Martin2112) [inline merkle/map_hasher.go]: Does this line up with the E2E map hashing? IIRC he's using tree level explicitly somewhere.

File: merkle/map_hasher.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	TreeHasher
+	keyHasher  keyHashFunc
+	nullHashes []trillian.Hash
+}
+
+// NewMapHasher creates a new MapHasher based on the passed in hash function.
+func NewMapHasher(hasher trillian.Hasher) MapHasher {
```
Comment (Martin2112) [inline merkle/map_hasher.go]: Pass in treehasher directly or is that a weird API?

File: merkle/map_hasher.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+	th := NewTreeHasher(hasher)
+	return MapHasher{
+		TreeHasher: th,
```
Comment (Martin2112) [inline merkle/map_hasher.go]: Ditto

File: merkle/map_hasher_test.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+import (
+	"bytes"
+	"encoding/base64"
+	"testing"
+
+	"github.com/google/trillian"
+)
+
+const emptyMapRootB64 = "xmifEIEqCYCXbZUz2Dh1KCFmFZVn7DUVVxbBQTr1PWo="
```
Comment (Martin2112) [inline merkle/map_hasher_test.go]: Is this created by hand?

File: merkle/sparse_merkle_tree.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+func NewSparseMerkleTreeReader(rev int64, h MapHasher, tx storage.ReadOnlyTreeTX) *SparseMerkleTreeReader {
+	return &SparseMerkleTreeReader{
+		tx:           tx,
+		hasher:       h,
+		treeRevision: rev,
+	}
+}
+
+func NewSparseMerkleTreeWriter(rev int64, h MapHasher, tx storage.TreeTX) *SparseMerkleTreeWriter {
```
Comment (Martin2112) [inline merkle/sparse_merkle_tree.go]: Needs a doc comment.

File: merkle/sparse_merkle_tree.go [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+	"github.com/google/trillian/storage"
+)
+
+type SparseMerkleTreeReader struct {
+	tx           storage.ReadOnlyTreeTX
+	hasher       MapHasher
+	treeRevision int64
+}
+
+type SparseMerkleTreeWriter struct {
```
Comment (Martin2112) [inline merkle/sparse_merkle_tree.go]: Needs comment

File: merkle/sparse_merkle_tree.go [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+	tx           storage.TreeTX
+	hasher       MapHasher
+	treeRevision uint64
+}
+
+var (
+	NoSuchRevision = errors.New("no such revision")
```
Comment (Martin2112) [inline merkle/sparse_merkle_tree.go]: Ditto

File: merkle/sparse_merkle_tree.go [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+func NewSparseMerkleTreeWriter(rev int64, h MapHasher, tx storage.TreeTX) *SparseMerkleTreeWriter {
+	return &SparseMerkleTreeWriter{
+		tx:     tx,
+		hasher: h,
+	}
+}
+
+func (s SparseMerkleTreeReader) RootAtRevision(rev int64) (trillian.Hash, error) {
```
Comment (Martin2112) [inline merkle/sparse_merkle_tree.go]: I'll stop flagging these now but api needs documenting.

File: storage/mysql/create_lots_of_node_data.sql [hunk 9]
```diff
@@ -0,0 +1,22 @@
+insert into Trees(TreeId) VALUES (0);
```
Comment (Martin2112) [inline storage/mysql/create_lots_of_node_data.sql]: Is this file required for tests to run? Don't see it referenced.

**Comment by Martin2112** [comment]
Should maybe add the sparse merkle paper to the repo. Not sure it's worth doing all the ascii art but the go file doesn't explain how it works.

File: merkle/map_hasher.go (hunk 1 above)
Comment (AlCutter) [inline merkle/map_hasher.go]: Possibly not; I'm not sure whether/why that's there, but I'll leave a TODO - it might be we need a configurable `TreeHasher` for these differences.

File: merkle/map_hasher.go (hunk 2 above)
Comment (AlCutter) [inline merkle/map_hasher.go]: Yeah, that's probably going to be required to allow configurable Tree/Map Hashers.
Done.

File: merkle/map_hasher_test.go (hunk 4 above)
Comment (AlCutter) [inline merkle/map_hasher_test.go]: Snarfed from another implementation :)

Comment (Martin2112) [inline merkle/map_hasher_test.go]: I've started adding comments on how to generate them so people can double check the value.

**Comment by Martin2112** [comment]
Did you push your changes?

File: storage/mysql/create_lots_of_node_data.sql (hunk 9 above)
Comment (AlCutter) [inline storage/mysql/create_lots_of_node_data.sql]: Whoops, no - that shouldn't be in there yet

**Comment by AlCutter** [comment]
I hadn't as I got interrupted by something, but have now pushed.

**Comment by Martin2112** [comment]
Ready to review again?

**Comment by AlCutter** [comment]
yup, PTAL
